### DISCUSSION TITLE
[Building pharos] error:  __is_cpp17_move_insertable<std::allocator<pharos::ParameterDefinition>, void>::value'

### DISCUSSION THREAD

**Discussion by didier31** [cmu-sei/pharos#252] [Q&A]
Hi,

I'm trying to build pharos and build failed due to requirement '__is_cpp17_move_insertable<std::allocator<pharos::ParameterDefinition>, void>::value'.

I guess some of you have encountered this error and could have relevant comment to do.

Thanks a lot.

_[  7%] Building CXX object libpharos/CMakeFiles/pharos.dir/convention.cpp.o
cd /home/didier/labo/pharos-release/libpharos && /usr/bin/c++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CHRONO_DYN_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_TIMER_DYN_LINK -DBOOST_TIMER_NO_LIB -DBOOST_WAVE_DYN_LINK -DBOOST_WAVE_NO_LIB -Dpharos_EXPORTS -I/home/didier/labo/pharos-release/libpharos -isystem /usr/local/include -isystem /home/didier/rose/usr/local/include/rose -isystem /home/didier/rose/usr/local/include -isystem /usr/local/lib/swipl/include -pedantic -pedantic-errors -Wall -Wextra -Wshadow -Wstrict-aliasing -Wno-misleading-indentation -ftemplate-depth=1024 -O3 -DNDEBUG -fPIC -pthread -std=c++14 -Winvalid-pch -Xclang -include-pch -Xclang /home/didier/labo/pharos-release/libpharos/CMakeFiles/pharos.dir/cmake_pch.hxx.pch -Xclang -include -Xclang /home/didier/labo/pharos-release/libpharos/CMakeFiles/pharos.dir/cmake_pch.hxx -MD -MT libpharos/CMakeFiles/pharos.dir/convention.cpp.o -MF CMakeFiles/pharos.dir/convention.cpp.o.d -o CMakeFiles/pharos.dir/convention.cpp.o -c /home/didier/git/pharos/libpharos/convention.cpp
In file included from <built-in>:410:
In file included from /home/didier/labo/pharos-release/libpharos/CMakeFiles/pharos.dir/cmake_pch.hxx:5:
In file included from /home/didier/rose/usr/local/include/rose/rose.h:10:
In file included from /home/didier/rose/usr/local/include/rose/sage3basic.hhh:4:
In file included from /home/didier/rose/usr/local/include/rose/sage3basic.h:42:
In file included from /usr/local/include/boost/filesystem.hpp:16:
In file included from /usr/local/include/boost/filesystem/path.hpp:21:
In file included from /usr/local/include/boost/filesystem/path_traits.hpp:14:
In file included from /usr/local/include/boost/system/error_category.hpp:10:
In file included from /usr/local/include/boost/system/detail/error_category.hpp:16:
In file included from /usr/include/c++/v1/string:522:
In file included from /usr/include/c++/v1/algorithm:653:
In file included from /usr/include/c++/v1/functional:500:
In file included from /usr/include/c++/v1/__functional/function.h:22:
/usr/include/c++/v1/memory:775:5: error: static_assert failed due to requirement '__is_cpp17_move_insertable<std::allocator<pharos::ParameterDefinition>, void>::value' "The specified type does not meet the requirements of Cpp17MoveInsertable"
    static_assert(__is_cpp17_move_insertable<_Alloc>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/vector:976:12: note: in instantiation of function template specialization 'std::__construct_backward_with_exception_guarantees<std::allocator<pharos::ParameterDefinition>, pharos::ParameterDefinition *>' requested here
    _VSTD::__construct_backward_with_exception_guarantees(this->__alloc(), this->__begin_, this->__end_, __v.__begin_);
           ^
/usr/include/c++/v1/vector:1691:5: note: in instantiation of member function 'std::vector<pharos::ParameterDefinition>::__swap_out_circular_buffer' requested here
    __swap_out_circular_buffer(__v);
    ^
/usr/include/c++/v1/vector:1709:9: note: in instantiation of function template specialization 'std::vector<pharos::ParameterDefinition>::__emplace_back_slow_path<unsigned long &, Sawyer::SharedPointer<pharos::SymbolicValue> &, char const (&)[1], char const (&)[1], nullptr_t, Sawyer::SharedPointer<pharos::SymbolicValue> &, unsigned long &>' requested here
        __emplace_back_slow_path(_VSTD::forward<_Args>(__args)...);
        ^
/home/didier/git/pharos/libpharos/convention.cpp:936:14: note: in instantiation of function template specialization 'std::vector<pharos::ParameterDefinition>::emplace_back<unsigned long &, Sawyer::SharedPointer<pharos::SymbolicValue> &, char const (&)[1], char const (&)[1], nullptr_t, Sawyer::SharedPointer<pharos::SymbolicValue> &, unsigned long &>' requested here
    d.params.emplace_back(num, invalid, "", "", nullptr, invalid, sd);_

**Comment by sei-mwd** [comment]
I can't say that I've run into this before.  Do you know what compiler and compiler version you are using? (`/usr/bin/c++ --version`, most likely.)

**Reply by sei-mwd** [reply]
Asa followup question: do you know what version of boost you are using?

**Reply by didier31** [reply]
Ah yes, it's boost 1.8

**Reply by sei-mwd** [reply]
I'm going to assume you meant boost 1.80, as 1.8 would be very, very old indeed.

**Comment by didier31** [comment]
Thank you a lot for your quick answer.
_$ c++ --version
FreeBSD clang version 13.0.0_

Infortunately, I'm not a C++ gentleman.

**Reply by sei-mwd** [reply]
Hmm...  I'll try to test with that version of clang.  We generally compile with gcc rather than clang, so if that's an option for you, you might want to try that instead.

**Reply by didier31** [reply]
OK, I'm switching to gcc.
Thank you again.
Have a nice day !

**Reply by didier31** [reply]
In fact, it was quite silly.

  ParameterDefinition(ParameterDefinition &&) = delete;
  ParameterDefinition &operator=(ParameterDefinition &&) = delete;

should be changed to :

  ParameterDefinition(ParameterDefinition &&) = default;
  ParameterDefinition &operator=(ParameterDefinition &&) = default;

**Reply by sei-mwd** [reply]
That may work in your specific circumstance, but we can't do that generally.  `ParameterDefinition` objects contain a mutex which is not defined to be movable by the STL.

**Reply by sei-mwd** [reply]
I was, unfortunately, unable to recreate this problem using clang++ v13.0.1 under Linux.

**Reply by sei-mwd** [reply]
Checking the standard more carefully, the fact that `ParameterDefinition`s are copyable should satisfy _Cpp17MoveInsertable_.  This appears to be a bug in the STL implementation for the specific compiler you are using.  Could you try using a more recent version of clang?

**Reply by didier31** [reply]
Cool ! Very nice from you to consider my trouble. I'm going to build again with clang15. Meanwhile, I've run the tests : 138/140 tests are OK.
